pkg/portfwd/client.go

```diff
@@ -24,26 +22,23 @@ func HandleTCPConnection(ctx context.Context, client *guestagentclient.GuestAgen
 		return
 	}
 
-	g, _ := errgroup.WithContext(ctx)
-
-	rw := &GrpcClientRW{stream: stream, id: id, addr: guestAddr}
-	g.Go(func() error {
-		_, err := io.Copy(rw, conn)
-		return err
-	})
-	g.Go(func() error {
-		_, err := io.Copy(conn, rw)
-		return err
-	})
+	//Handshake message to start tunnel
+	err = stream.Send(&api.TunnelMessage{Id: id, Protocol: "tcp", GuestAddr: guestAddr})
+	if err != nil {
+		logrus.Errorf("could not start tcp tunnel for id: %s error:%v", id, err)
+		return
+	}
 
-	if err := g.Wait(); err != nil {
-		logrus.Debugf("error in tcp tunnel for id: %s error:%v", id, err)
+	remote := tcpproxy.DialProxy{
+		DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+			rw := &GrpcClientRW{stream: stream, id: id, addr: guestAddr, protocol: "tcp"}
+			return rw, nil
+		},
 	}
+	remote.HandleConn(conn)
```

Comment:
Delegating the code to another libary is nice, if the library we used is reliable and well maintained.

Is this same tcpproxy libary that does not log anything on errors and was replaced in vz/network_darwin.go?

We must have 2 properties:
- when one side is closed or fail during a copy, the other side is closed
- when the first copy fail, the error is logged or returned to the caller

Example failure flow:
1. first io.Copy() blocked reading from host side socket
2. second io.Copy() blocked reading from grpc socket
3. host side socket closed, first io.Copy() return
4. second io.Copy() continue to block reading from grpc
5. no error is logged since error handler is waiting for both io.Copy to finish

If this library does not ensure both properties we should not use it. It will impossible to debug when we have issues.

We can mitigate the logging issues by logging errors in Read() and Write() since forwarding will stop on the first error.

Reply:
Looking in current tcpproxy used, it is the same library with the same issues:

```go
	errc := make(chan error, 1)
	go proxyCopy(errc, src, dst)
	go proxyCopy(errc, dst, src)
	<-errc
```

The errors from io.Copy are dropped silently.

```go
func proxyCopy(errc chan<- error, dst, src net.Conn) {
	// Before we unwrap src and/or dst, copy any buffered data.
	if wc, ok := src.(*Conn); ok && len(wc.Peeked) > 0 {
		if _, err := dst.Write(wc.Peeked); err != nil {
			errc <- err
			return
		}
		wc.Peeked = nil
	}

	// Unwrap the src and dst from *Conn to *net.TCPConn so Go
	// 1.11's splice optimization kicks in.
	src = UnderlyingConn(src)
	dst = UnderlyingConn(dst)

	_, err := io.Copy(dst, src)
	errc <- err
}
```

When one io.Copy finish (success of failure), the other io.Copy is not affected. If both are blocked on Read() or Recv(), what would cause the call to return?

We have the same issue in the current code, but at least in the current code we will log the error if both io.Copy() fail.

It can be useful to fix both issues in tcpproxy but it will be easier to fix them in lima now. Is this the same code used for user v2 network?

Reply:
I want to keep the port forwarding stuffs consistent across portfwclient, portfwserver, user-v2, vz forwardings.

In server we need to use tcpproxy because we need to dial and pass connection. So i wanted to use the same on client side as well.

Additionally, gvisor-tap-vsock userspace network is already using these for proxy we are using the same for vz and user-v2 network. This library has been reliable from that point of view.

I intend to reuse as these are testing across different products (not just lima), so that we can avoid failures with custom implementation like before.

Reply:
Which file you mean ???

Reply:
Updated to use bicopy for server and client (Only tcp connection).

Upstream copied tcpproxy because they had a connection leak issue in latest tcpproxy version

Reply:
> Upstream copied tcpproxy because they had a connection leak issue in latest tcpproxy version

Did you report that issue to upstream? Where is the fix?

Reply:
Check the git log for such reference

https://github.com/containers/gvisor-tap-vsock/commit/c315e611503f22d0b27f5c9f2857f823c3aff325

FYI - https://github.com/containers/gvisor-tap-vsock/issues/387

Reply:
@tamird Am the one shared those links so i know what i shared. 

Marking this as resolved as its not relevant to us we are using bicopy only already